### ISSUE TITLE
Add support for API documentation generation

### ISSUE THREAD

**Issue by agoose77** [executablebooks/mystmd#1259]
## Proposal

We would like to be able to use MyST for API documentation. It's already very close to possible; adding support for declaring functions, classes, etc, just requires some new directives and some labelling logic. To support more advanced documentation tools, like `sphinx-autodoc`, we could leverage Sphinx itself.

## Related issues:
- https://github.com/executablebooks/mystmd/issues/293

**Comment by choldgraf** [comment]
Here's a related issue that is specifically about docstrings https://github.com/jupyter-book/mystmd/issues/1858

**Comment by rowanc1** [comment]
From @fwkoch on Discord:

> This is a demo of some work @stefanv and I did at the Scientific Python Developer Summit. The goal was to get autogenerated Python docs rendering in MyST, a critical step to enable websites with technical API docs to upgrade to MyST. Let me know if you have any thoughts or questions!
> https://www.loom.com/share/589a947389fb49829c9b712b796d8a7d

**Comment by echarles** [comment]
> This is a demo of some work @stefanv and I did at the Scientific Python Developer Summit.

Looks awesome! Is there a link to a repo with the source of that work?

**Comment by rowanc1** [comment]
The Python code and the MyST plugin are here: https://github.com/stefanv/npdoc2json

**Comment by romanlutz** [comment]
@rowanc1 any plans for Google-style docstrings?

From a cursory glance, it seems like it may be as simple as just replacing the Python script with logic to parse the Google-style docstrings. I might give this a try one of these days and report back 🙂

**Comment by atharvaaalok** [comment]
Has there been any progress on generating API documentation using docstrings?
If so, is there any resource/webpage that I could use to create my own MyST based API docs?
Cheers!

**Comment by stefanv** [comment]
@romanlutz Yeah, exactly! The beauty of this demonstrator is that it is pretty easy to modify to work with any format, as long as you can parse it. I will put a few more cycles into this soon, perhaps try it on scikit-image or similar.

**Comment by atharvaaalok** [comment]
@stefanv Is there anything planned for incorporating the functionality of building documentation through an API call in the near future as a proper feature in MyST?

**Comment by julian-belina** [comment]
Are there any plans to make this possible? V2 looks much nicer, but without the ability to automatically generate API documentation, it's not suitable for my use cases.

**Comment by choldgraf** [comment]
Right now the "complex" project the team is tackling is the PRs to generate AST directly from code cells. After that, I think the next focus will likely be a widget-like interface to allow for more flexible JavaScript use by plugins.

After that I don't see an obvious "big" project driving the community, so maybe something like this will come up then. As a reminder there are no dedicated resources to the project so things move more slowly than having dedicated development capacity.

**Comment by thangckt** [comment]
I am also looking forward to this feature.
Hope to see it soon.

Any update so far?

Thanks.

**Comment by sbonaretti** [comment]
Thanks for all your work! I hope to see this soon too! :)

**Comment by danielballan** [comment]
I'm commenting to add myself and ~~@tomcobb~~ @coretl to this conversation. We are part of an open-source collaboration between X-ray user facilities, and that collaboration may have some bandwidth to contribute in this area.

In a conversation on the Discord, @stefanv shared his work on a [prototype](https://github.com/stefanv/myst-apidoc) and a [preview](https://circle.scientific-python.dev/output/job/0d5f40cf-a25a-4055-acb4-c89106271094/artifacts/0/_build/html/skimage). He summarized its operation:

>Docstring extraction -> JSON (currently a script iterating over functions and running numpydoc to extract parts, we couldn't get the info we need out of griffe)
> Post-process JSON (convert RST descriptions to markdown)
> JSON -> myst AST

**Comment by danielballan** [comment]
Perhaps a good next step here would be for some of us to join one of your working sessions and see if we can get some of our documentation running on the prototype, and identity any gaps or action items that follow from that.

**Comment by stefanv** [comment]
That'd be great! If you share the name of some packages beforehand, I can make sure the docstring extractor runs on them.

**Comment by coretl** [comment]
> I'm commenting to add myself and [@tomcobb](https://github.com/tomcobb) to this conversation. We are part of an open-source collaboration between X-ray user facilities, and that collaboration may have some bandwidth to contribute in this area.

Apologies to @tomcobb, @danielballan meant to tag me instead...

> That'd be great! If you share the name of some packages beforehand, I can make sure the docstring extractor runs on them.

The repo I was thinking about converting is https://github.com/bluesky/ophyd-async. It currently uses [sphinx-autodoc2](https://github.com/sphinx-extensions2/sphinx-autodoc2) to generate the API docs, which requires the use of sphinx style docstrings, but we have used google style and numpy style docstrings in the past. The docstrings and accompanying prose are all written in myst markdown.

**Comment by danielballan** [comment]
That seems like a clean one to start with. I would think about doing https://github.com/bluesky/tiled next.

**Comment by stefanv** [comment]
Thanks, @coretl. I read through the docs for sphinx-autodoc2, and I cannot find a definition of the format; do you mind pointing me to it?

**Comment by stefanv** [comment]
@danielballan Looks like `tiled` uses numpydoc (with [some exceptions](https://github.com/bluesky/tiled/blob/main/tiled/client/cache.py#L24)), so that one should be easy.

**Comment by stefanv** [comment]
I think I found the [Sphinx docstring format](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html), although not clear how the specification is altered exactly when mixing in markdown—presumably just for math, warning directives, etc.

**Comment by coretl** [comment]
> Thanks, [@coretl](https://github.com/coretl). I read through the docs for sphinx-autodoc2, and I cannot find a definition of the format; do you mind pointing me to it?

There is an example in the tip on this page:
https://sphinx-autodoc2.readthedocs.io/en/latest/quickstart.html#using-markdown-myst-docstrings


> I think I found the [Sphinx docstring format](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html), although not clear how the specification is altered exactly when mixing in markdown—presumably just for math, warning directives, etc.

I'm not sure either, I've only ever put the param lists directly after the first sentence of docstring, outside any other directive

**Comment by danielballan** [comment]
This is somewhat related, as we have based some repos (e.g., [bluesky-tiled-plugins](https://github.com/bluesky/bluesky-tiled-plugins)) off of this template. https://github.com/scientific-python/cookie/pull/792

**Comment by coretl** [comment]
I've done a bit of prototyping work, and I think I understand enough to ask some informed questions here.

Our aim is to migrate our sphinx + myst-parser + sphinx-autodoc/sphinx-autodoc2 stack to mystmd + myst-apidoc. We wanted to enable interop as best we could, both to and from other sphinx and mystmd projects. #1758 and #2988 enable case sensitivity on the consuming side (so both `re.match` and `re.Match` can be different targets both locally and via intersphinx). The remaining gap is on the producing side, and specifically what is published in `objects.inv`.

At the moment only `std:doc`/`std:label` are published, so downstream sphinx projects that reference `py:class` for instance would not resolve. I could think of a couple of options:

1. Live with it, only emit `std:label` rows and accept some changes in downstream sphinx projects
2. Add a convention on kind, so that it can contain the domain and objType, e.g. `py:class`, and write it into the rows in the inventory writer (mentioned by @fwkoch in #2940)

If we went with 2, then kind would appear in `myst.xref.json`, and it would open up the possibility of a `{py}` role for myst consumers (e.g. `` {py}`re.match` ``)

When this question is resolved my plan is to fold it into our ([PoC](https://github.com/gilesknap/mystmd-pythonapi)) and then discuss our findings with @stefanv so we can compare and contrast with his approach in [myst-apidoc](https://github.com/stefanv/myst-apidoc) and hopefully work together on something that will meet all our needs!

Am I asking the right questions? Is there an agreed direction of travel in this area?
